File: Greedy/1717. Maximum Score From Removing Substrings/solution.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        def calculate(first):
            res = 0
            stack = []
            for ch in s:
                if not stack:
                    stack.append(ch)
                elif stack[-1] == first[0] and ch == first[1]: # ba
                    stack.pop()
                    res += (y if first == "ba" else x)
                else:
                    stack.append(ch)

            stack2 = []
            for ch in stack:
                if not stack2:
                    stack2.append(ch)
                elif stack2[-1] == first[1] and ch == first[0]: # ab
                    stack2.pop()
                    res += (x if first == "ba" else y)
                else:
                    stack2.append(ch)
            return res

        return max(calculate("ab"), calculate("ba"))

File: Greedy/1717. Maximum Score From Removing Substrings/solution.cpp
// ab -> ba
// ba -> ab
class Solution {
public:
    int check(string s, int x, int y, bool reverse) {
        string s1 = "ab", s2 = "ba";

        if (reverse) {
            swap(s1, s2);
            swap(x, y);
        }

        stack<char> st;
        int res = 0;
        for (char& c : s) {
            if (!st.empty() && st.top() == s1[0] && c == s1[1]) {
                st.pop();
                res += x;
            } else {
                st.push(c);
            }
        }

        stack<char> st2;
        while (!st.empty()) {
            char c = st.top();
            st.pop();

            if (!st2.empty() && st2.top() == s1[0] && c == s1[1]) {
                st2.pop();
                res += y;
            } else {
                st2.push(c);
            }
        }
        return res;
    }
    int maximumGain(string s, int x, int y) {
        return max(check(s, x, y, false), check(s, x, y, true));
    }
};

